source/async_stack.cpp

```diff
@@ -38,11 +41,48 @@ extern "C" {
 // Initialise to some value that will be interpreted as an invalid key.
 inline pthread_key_t folly_async_stack_root_tls_key = 0xFFFF'FFFFu;
 }
-
-#endif  // FOLLY_ASYNC_STACK_ROOT_USE_PTHREAD
+#endif  //UNIFEX_ASYNC_STACK_ROOT_USE_PTHREAD
 
 namespace unifex {
 
+#if UNIFEX_ASYNC_STACK_ROOT_USE_PTHREAD == 0
+#include <vector>
+
+struct AsyncStackRootHolderList {
+  std::vector<void*> asyncStackRootHolders_;
+  std::mutex mutex_;
+
+void add(void* holder) noexcept {
+  std::lock_guard<std::mutex> lock(mutex_);
+  asyncStackRootHolders_.push_back(holder);
+}
+
+void remove(void* holder) noexcept {
+  if (mutex_ == nullptr) {
+    return;
+  }
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto it = std::find(asyncStackRootHolders_.begin(), asyncStackRootHolders_.end(), holder);
+  if (it != asyncStackRootHolders_.end()) {
+    std::swap(*it, asyncStackRootHolders_.back());
+    asyncStackRootHolders_.pop_back();
+  }
+}
+
+std::vector<void*> getAsyncStackRoots() noexcept {
+  if (!asyncStackRootsMutex.try_lock()) {
+    // assume we crashed holding the lock and give up
+    return {};
+  }
+  std::lock_guard<std::mutex> lock(mutex_, std::adopt_lock);
+  return asyncStackRootHolders_;
+}
+
+extern "C" auto* kUnifexAsyncStackRootHolderList = new AsyncStackRootHolderList();
```

Comment:
don't we need to explicitly delete this to avoid leaks?

Reply:
We would, but leaking on purpose allows you to avoid worrying about background threads using this after the main thread has run after-main() global destructors.